win-capture: Add compatibility notices

derrod opened the pull request:
### Description

Adds text warnings with optional URLs to window and game capture to notify users about incompatibilities using the new text labels introduced in #5160.

The update mechanism for the compatibility data is the same as used in rtmp-services, meaning that it will fetch updated data from the obsproject website on startup.

**Screenshots:**
Window capture (warning):
![2023-02-05_17-26-21_mGUCRU](https://user-images.githubusercontent.com/3123295/216831602-4e1844d6-057c-4b38-be35-b0e848f66d23.png)

### Motivation and Context

Provide in-app guidance for users when things don't work in case of known compatibility issues.

### How Has This Been Tested?

Mucked about locally.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

derrod commented:
Based on off-chain feedback translation is a potential issue.

My idea would be:
- Add `translation_key` and `title` properties to JSON
- Provide translatable generic messages, like `%title% may require OBS to be run as admin to use Game Capture`
- Provide translatable specified message, e.g. for CSGO and Minecraft
- If the translation string exists, attempt to replace variables, i.e. `%title%` in string => `title` from JSON
- If no translation key exists, e.g. outdated OBS version or new entry where generic messages couldn't be used, use message from JSON instead

derrod commented:
Rebased on master with the following changes:

- Remove labels (as discussed off-chain yesterday)
- Added explanation of severity labels to schema
- Squashed commits

Gol-D-Ace commented:
Gave this a try and it looks like if no application is selected it defaults to the CSGO message for me.
![Screenshot_20230205-162221-489](https://user-images.githubusercontent.com/5058319/216828309-f12f8941-e2b9-4b6e-a20f-cd18ae985e84.png)

derrod commented:
Yeah I didn't consider the "any" case. I fixed that now (Also added GTA:SA and SAMP entries based on Discord discussion).

RytoEX commented on plugins/win-capture/data/locale/en-US.ini:
```diff
@@ -39,3 +39,13 @@ GameCapture.Rgb10a2Space="RGB10A2 Color Space"
+Compatibility.WindowCapture.BitBlt="%name% may not be capturable using the selected Capture Method (BitBlt)."
+
+# Specific compatibility messages
+Compatibility.Application.CSGO="CSGO may require the <code>--allow_third_party_software</code> launch option to use Game Capture"
```
```suggestion
Compatibility.Application.CSGO="CS:GO may require the <code>--allow_third_party_software</code> launch option to use Game Capture"
```

RytoEX commented on plugins/win-capture/data/locale/en-US.ini:
```diff
@@ -39,3 +39,13 @@ GameCapture.Rgb10a2Space="RGB10A2 Color Space"
+Compatibility.WindowCapture.BitBlt="%name% may not be capturable using the selected Capture Method (BitBlt)."
+
+# Specific compatibility messages
+Compatibility.Application.CSGO="CSGO may require the <code>--allow_third_party_software</code> launch option to use Game Capture"
+Compatibility.Application.Minecraft="If you're having issues capturing Minecraft: Java Edition check our troubleshooting guide"
```
Sentence ending punctuation?

RytoEX commented on plugins/win-capture/data/schema/compatibility-schema-v1.json:
```diff
@@ -0,0 +1,99 @@
+            "maximum": 2
+        }
+    }
+}
```
End with newline.

RytoEX commented on plugins/win-capture/game-capture.c:
```diff
@@ -2247,8 +2249,38 @@ static bool mode_callback(obs_properties_t *ppts, obs_property_t *p,
+	const char *window =
+		obs_data_get_string(settings, SETTING_CAPTURE_WINDOW);
+
+	char *class, *exe, *title;
```
Declare these on separate lines.

RytoEX commented on plugins/win-capture/plugin-main.c:
```diff
@@ -75,16 +107,30 @@ bool obs_module_load(void)
+				confirm_compat_file, NULL);
+		}
 	}
+	bfree(local_dir);
+	bfree(config_dir);
```
Free in reverse order of declaration.

derrod commented on .github/workflows/compat-json.yml:
```diff
@@ -0,0 +1,40 @@
+jobs:
+  schema:
+    name: Schema
+    runs-on: [ubuntu-20.04]
```
No specific reason, this was just copied 1:1 from the rtmp-service schema checker.

derrod commented on plugins/win-capture/data/compatibility.json:
```diff
@@ -0,0 +1,174 @@
+            "game_capture": true,
+            "window_capture": false,
+            "window_capture_wgc": false,
+            "message": "Valorant may require OBS to be run as admin to use Game Capture.",
```
I do not have that game installed, I just added entries from https://obsproject.com/kb/game-capture-troubleshooting

derrod commented on plugins/win-capture/plugin-main.c:
```diff
@@ -1,8 +1,18 @@
+#include "compat-format-ver.h"
+#include "compat-config.h"
+
+#define WIN_CAPTURE_LOG_STRING "[win-capture plugin] "
```
This is just a straight copy from `rtmp-services`:

https://github.com/obsproject/obs-studio/blob/master/plugins/rtmp-services/rtmp-services-main.c#L21-L22

An example line would be
`20:47:45.242: [win-capture plugin] Remote update of URL "https://obsproject.com/obs2_update/win-capture/v1/package.json" failed: The requested URL returned error: 404`

RytoEX replied:
Weird. If it's only used in that one place and is based on an existing pattern, then I'm not as concerned.

derrod commented:
Based on Flaeri's feedback:
- Removed Valorant (seems to be no longer reuqired, possibly never was)
- Added generic Electron entries to BitBlt Window Capture and Game Capture
  + Window Capture will default to WGC for Electron apps, so only a "notice" severity
  + For Game Capture this affects titles such as Vampire Survivors that are built using Electron. See https://steamdb.info/tech/SDK/Greenworks/ for a list of many affected titles

RytoEX reviewed (approved):
Looks okay to me. I think this is a good first attempt at trying to give users some assistance to resolve issues on their own without having to come to our support channels.
